_posts/2025-05-09-sorting-for-fast-selective-queries.md

```diff
@@ -0,0 +1,660 @@
+
+Our hypothetical use case is to serve a dashboard where people can explore a single `origin`, a single `dest` (destination), or an `origin`–`dest` pair.
+The `origin` and `dest` columns are three-letter airport codes (like `LAX` or `JFK`).
+We assume once you have picked one of those filters, you want all the rows that match and a few arbitrary columns.
+
+Our second experiment will add in a time component and additionally filter to the latest 1, 13, or 52 weeks of data in the dataset.
+
+## Experimental Results
+
+### Multi-Column Sorting
+
+We will show a few approaches to sort data approximately or using space-filling curves to speed up retrieval.
+Let's have a look at some illustrative results first and dive into the details in subsequent sections.
+
+The control groups are:
+
+- `random`: Sort by a hash – the worst case scenario!
+- `origin`: Single-column sort on `origin`
+- `origin-dest`: Sort on `origin`, then on `destination`
+
+Our alternative sorting approaches are:
+
+- `zipped_varchar`: Sort by one letter at a time, alternating between `origin` and `destination`. This approach is implemented using a custom SQL macro – see the [Appendix](#appendix-experiment-details) for details!
+- `morton`: Convert `origin` and `destination` into integers, then order by [Morton encoding (Z-order)](https://en.wikipedia.org/wiki/Z-order_curve)
+- `hilbert`: Convert `origin` and `destination` into integers, then order by [Hilbert encoding](https://en.wikipedia.org/wiki/Hilbert_curve)
+
+> The Morton and Hilbert encoding functions come from the [`lindel` DuckDB community extension]({% link community_extensions/extensions/lindel.md %}), contributed by [Rusty Conover](https://github.com/rustyconover).
+> Thank you to Rusty and the folks who have built the [`lindel` Rust crate](https://crates.io/crates/lindel) upon which the DuckDB extension is based!
+> The [`spatial` extension]({% link docs/stable/extensions/spatial/overview.md %}) also contains an [`ST_Hilbert` function]({% link docs/stable/extensions/spatial/functions.md %}#st_hilbert) that works similarly.
+> Thanks to [Max Gabrielsson](https://github.com/Maxxen) and the GDAL community!
+
+Both Morton and Hilbert are space filling curve algorithms that are designed to combine multiple columns into an order that preserves some approximate ordering for both columns.
+In the geospatial world, they are often used to preserve approximate ordering of latitude and longitude.
+They operate on integers or floats, so a SQL macro is used to convert the first few characters of a `VARCHAR` column into an integer as a pre-processing step.
+More details are included in the [Appendix](#appendix-experiment-details)!
+
+These plots display query runtime when pulling from a DuckDB file hosted on S3.
+Querying a local DuckDB file showed very similar patterns, but with much faster query times overall!
```

Comment:
We may want to skip this. IMHO the key point of the techniques in the blog post is speeding up remote access.

Reply:
I think I will reword it a bit to say that remote file access makes this a larger optimization. I think that will balance it!

Reply:
Happy to change back, but I took a stab at rewording!